### DISCUSSION TITLE
What hardware is required?

### DISCUSSION THREAD

**Discussion by EUTW06** [OpenRLHF/OpenRLHF#250] [Q&A]
What is the minimum hardware (GPU, VRAM) to train if I want using LoRA for SFT and KTO instead of fulltuning? (7B)
Thank you

**Comment by gyanu2507** [comment]
For a 7B with LoRA SFT, a single 24 GB GPU (4090 / A5000 / L4 24GB) is the comfortable floor. 16 GB (T4 16GB, 4060 Ti 16GB) works with 4-bit base + LoRA, sequence length 2k, small batch, gradient checkpointing. 12 GB is possible but you will fight context length.

KTO is closer to DPO than to SFT. You need policy and a reference. With LoRA, the ref can be the frozen base (no second full copy), but activations for two forwards still land in VRAM. Budget ~1.5x to 2x the SFT setup. 24 GB is the realistic minimum for 7B KTO+LoRA at 2k context. 2x 24 GB if you want 4k+ or a larger batch.

Full 7B SFT without LoRA is a different class (roughly 60 GB+ fp16, or ZeRO-3 across several 24 GB cards). You said you do not want that.

Start with one 24 GB GPU, LoRA rank 16-64, QLoRA 4-bit base for KTO if the first run OOMs. NVLink is not required at this size.
